protocols/kad/src/handler.rs

```diff
@@ -766,15 +777,31 @@ where
             ));
         }
 
-        if let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
+        let num_in_progress_outbound_substreams =
+            self.active_outbound_substreams.len() + self.num_upgrading_outbound_streams;
```

Comment:
```suggestion
            self.active_outbound_substreams.len() + self.next_outbound_streams.len();
```

can't we just use `next_outbound_streams` here? why do we need a separate counter?

Reply:
If `active_outbound_substreams` contains `ConnectionHandlerEvent::OutboundSubstreamRequest`, it'll be returned from `poll` and eventually from `active_outbound_substreams` too (though not immediately). It'll be later put back into `active_outbound_substreams` as `OutboundSubstreamState::PendingSend` or `OutboundSubstreamState::ReportError`.

So the counter is used to track those streams that temporarily not in `active_outbound_substreams` and are pending re-insertion with an actual substream allocated to it.

`num_in_progress_outbound_substreams` holds number of all streams (both in `active_outbound_substreams` and temporarily not there) and once the sum goes below `MAX_NUM_SUBSTREAMS` we pull one more stream to work on from `next_outbound_streams`.

So the refactoring suggested doesn't really make much sense and will break the intended logic, hopefully above explanation makes it clearer as to why.